Package version data

kostiakoval opened the pull request:
**Work in Progress** 
This is initial commit for [SR-473](https://bugs.swift.org/browse/SR-473)

We need to generate version data for every package on build.
Now it creates a new folder `versionData` inside `.build` dir and creates a swift file for every package.

Could you please provide any feedback for the initial implementation. 
- Should these files be inside `.build` dir or somewhere else?
- Should we generate a new `.swift` file for every Package or place all data in 1 file?
- Any other suggestion?

ToDo:
- [x] Add unit test
- [x] Add better error handling
- [x] Merge PR #225 
- [x] add additional information to Root package (local changes)

mxcl commented:
Ideally these would be targets integrated into the build yaml so they are not generated every time. However not doing it this way at first is acceptable since we don't have a good way to specify the version dependency yet.

aciidb0mb3r commented on Sources/swift-build/versionData.swift:
```diff
@@ -0,0 +1,36 @@
+
+    try fputs("public let url: String = \"\(package.url)\" \n", file)
+    let version = package.version
+    try fputs("public let version: (Int, Int, Int, [String], String?) = " +
```
I think this would be better if tuple had named values?

kostiakoval replied:
@aciidb0mb3r something like: 
```swift 
 public let version: (major: Int, minor: Int, patch: Int, etc: [String], metadata: String?)
```

aciidb0mb3r replied:
yup

kostiakoval replied:
I think using type inference would be even better:
What do you think @aciidb0mb3r @mxcl ?
```swift
public let version: (major: Int, minor: Int, patch: Int, etc: [String], metadata: String?) = (1, 0, 1, [], "")
//vs
public let version = (1, 0, 1, [], "")
```

aciidb0mb3r replied:
imo, it would be fine if it was just `public let version = (1, 0, 1)` as this is guessable but I have no way of knowing what is last two fields in `public let version = (1, 0, 1, [], "")` without seeing doc

kostiakoval replied:
It's [Version](https://github.com/apple/swift-package-manager/blob/master/Sources/PackageDescription/Version.swift) properties.
-  prereleaseIdentifiers: [String]
-  buildMetadataIdentifier: String?
They describe a version tag of a package. So we need them.

mxcl commented:
Let’s resurrect this potentially fabulous addition.

kostiakoval commented:
I'm sorry for being lazy here. I'm going to update it soon

mxcl commented:
“Lazy” is certainly not the right word. But we owe it to everyone to prod each other and keep momentum up :)

kostiakoval commented:
Hi all. I'm back :) 

I'm thinking of adding `version: Version?` property to a `Package`type. It's optional because local package could be missing a git version.
What do you think @mxcl?

jessesquires commented:
@kostiakoval an optional seems like a bad idea to me.

@mxcl isn't this something that SPM intends to enforce? 

i'm not sure how effect a package manager could be without version numbers :smile:

kostiakoval commented:
Hi @jessesquires I'm not talking about a Version you specify in your Manifest file when you describe a package, that must be specified and it's not optional.

The `Version` I described, is a version we get when we fetch a package.

Local packages `.Package(url: "../Foo", versions: Version(1,0,0)..<Version(2,0,0))`
It could have missing version tag, and that should be fine when you are developing a package (not published yet).  
This PR, represents a metadata (url and Version) for every package swiftpm is fetching and building.

Does it make any sense, any suggestion how make it better?

jessesquires commented:
Ah! :blush: I see. In that case, we probably do want an optional.

Maybe we could rename to `versionTag` or something similar to avoid confusion?

mxcl commented:
No optional, I think in the case of the root package where we can't get a version just go with `0,0,0`. It would suck to always check for the optional here and need to code an error path when it is in fact a programmer error,

The other option is `Version!`

mxcl commented:
Unless this is only being exposed to SwiftPM, which seems likely, so `Version?` is ok in that case.

kostiakoval commented:
Done. Review it please.

On build generate swift file for every package with Version data.
Each file contains:
```swift
public let url: String
public let version: (Int, Int, Int, [String], String?)?
```

kostiakoval commented:
@mxcl Can you please run CI and merge it?

aciidb0mb3r commented on Sources/Build/PackageVersion.swift:
```diff
@@ -0,0 +1,50 @@
+import func libc.fclose
+
+public func generateVersionData(rootDir: String, packages: [Package]) throws {
+    let dirPath = rootDir + "/.build/versionData/"
```
prefer `Path.join`

aciidb0mb3r commented on Sources/Build/PackageVersion.swift:
```diff
@@ -0,0 +1,50 @@
+    try mkdir(dirPath)
+    let versionData = generateData(packages)
+
+    for (pkgName, data) in versionData {
```
don't really need local var `versionData`

aciidb0mb3r commented on Sources/Build/PackageVersion.swift:
```diff
@@ -0,0 +1,50 @@
+}
+
+func generateData(packages: [Package]) -> [String : String] {
+    var versionData: [String : String] = [:]
```
I wonder if `var versionData = [String : String]()` is more concise here?

aciidb0mb3r commented on Tests/Build/Utilities.swift:
```diff
@@ -0,0 +1 @@
+../Functional/Utilities.swift
```
not sure what is this?

kostiakoval replied:
That' s a link to share `Utilities.swift` file. It has `mktmpdir` and other utility functions.

aciidb0mb3r replied:
a symlink? okay din know it shows up in git like this

kostiakoval replied:
yes it does. 
https://github.com/apple/swift-package-manager/blob/master/Tests/ManifestParser/Utilities.swift

kostiakoval commented:
@aciidb0mb3r  Thanks for feedback. The issue you posted out are fixed now :+1:

mxcl commented:
I've changed my view here.

If root package version data cannot be obtained, do not generate the version swift.

If it can be obtained but commits or work has been done *since* a version, then add a build identifier with the date.

kostiakoval commented:
@mxcl Can we review and merge this PR? Some parts of it can be reused when implementing `PackageVersions.json` file

kostiakoval commented:
I've added modified date for a root packages if it contains changes. Here is a generated file
```swift
public let url: String = "/Users/kkoval/Work/apple/swiftpm" 
public let version: (Int, Int, Int, [String], String?)? = (0, 2, 1, [], nil) 
public let sha: String = "d9a6a39f85a3d3c6dcbac38f379692370d1baca1" 
public let modified: String = "2016-04-09T16:50:37Z" 
```
@mxcl Do you have any feedback? it is ready for merging

ddunbar commented:
How is a client going to use modified if it isn't only present if there are local changes? Should it be an optional and nil if there are no changes?

kostiakoval commented:
@ddunbar Now the values are options and variables alway present for clients.

ddunbar commented on Sources/Build/PackageVersion.swift:
```diff
@@ -33,13 +33,17 @@ func saveRootPackage(_ dirPath: String, package: Package) throws {
+        let date = try strftime("%Y-%m-%dT%XZ", time: gmtime_r())
+        data += " = \"\(date)\""
     }
+    data += "\n"
```
Isn't this missing a "= nil" in the non-local changes case?

kostiakoval replied:
Optionals are nil by default. I thought `public let modified: String?` would be enough.

ddunbar replied:
You are right, I hadn't realized that yet! Thanks :)

mxcl commented:
Oh sweet. I'll check it out today.

aciidb0mb3r commented:
`public let version: (Int, Int, Int, [String], String?)? = (0, 2, 1, [], nil)`
I am still against this, there is no way of guessing what is `[String]` and `String?`

`public let modified: String = "2016-04-09T16:50:37Z" `
unix timestamp is better IMO

mxcl commented:
I agree, we should name the last two:

```
public let version: (Int, Int, Int, prereleaseIdentifiers: [String], buildMetadata: String?)? = (0, 2, 1, [], nil)
```

kostiakoval commented:
Agree, It should the same as `Version` init format.
Changed.

ddunbar commented:
If the last two are being named, would it be worth it to just name all of them? It would probably be more readable, and I can't really see a strong downside.

kostiakoval commented:
I'm ok with adding names for all values, like:
```swift 
public let version: (major: Int, minor: Int, patch: Int, prereleaseIdentifiers: [String], buildMetadata: String?)? = (0, 2, 1, [], nil)
```

 My motivation was to keep it consistent with `Version` init method
```swift
Version(1,2,3))
Version(1,2,3, prereleaseIdentifiers: ["alpha", "beta"], buildMetadataIdentifier: "1011"))
```

Should I add names?

ddunbar commented:
Up to Max...

mxcl commented:
I still think these should not be optional, instead just don't generate them if they don't exist.

Rationale: this is compile time functionality, not runtime functionality.

Otherwise this looks good. I notice it isn't possible to use the data yet, so that could be an easy next PR.

ddunbar commented:
Maybe I don't understand, how would other code use them if they aren't optional?

mxcl commented:
I think you understand: other code can't use them if they don't exist.

Thus they won't exist:

1. For new packages not yet ever published
2. Projects that never intend to have versions

I think it is reasonable for version information to not exist when it literally doesn't exist.

mxcl commented:
K, talked with Daniel and we think:

1. `version` should not be optional, if there is no version, don't generate it. We think this is reasonable so that: people don't have to add error handling for when the version doesn't exist considering the times it doesn't exist are very edge case
2. We agree she should remain optional and we like how it works
3. We think the modified string should be changed to a single `Bool`. Our thinking here is that we don't think it the build-time has much utility in the context of this PR and we really don't want people writing code that does things based on this date value.
4. Let's label all of the tuple values in version.
5. Let’s add a `versionString` property that returns a string since mostly this is a feature for logging versions

Additional longer term considerations:

1. We think that this should *only* be generated for the root-package, the rationale here is that we don't want packages writing code like `if version.major > 2` as this should be a compile time language feature and Swift core are already on board with this idea.
2. Thus we think we should generate a single module called `PackageVersions` (depending on the version-pinning proposal’s end name)
3. And thus we should probably define a struct for each package in that file

I'll let Daniel correct me if he has any corrections. Otherwise, thoughts?

ddunbar commented:
I thought we agreed for assigning `let version = (0,0,0,...)` if missing -- I think this is better so that you can test the code which would use it before you have tagged. Otherwise you have to tag, build, test, retag.

Otherwise yup that matches my thoughts.

One thing to note about the implementation is you will need to treat PackageVersions as a dependency of every target in the top-level project (so they can import it, potentially).

mxcl commented:
> I thought we agreed for assigning let version = (0,0,0,...) if missing -- I think this is better so that you can test the code which would use it before you have tagged. Otherwise you have to tag, build, test, retag.

Oh yes, sorry. You are right. I agree.

mxcl commented:
> One thing to note about the implementation is you will need to treat PackageVersions as a dependency of every target in the top-level project (so they can import it, potentially).

Indeed. Though @kostiakoval it is cool if we do this part of the implementation in a separate PR.

kostiakoval commented:
> We think that this should only be generated for the root-package, 

Can you please clarify what you mean by `root-package`

It's a package with `Package.swift` file and it's always only 1. For `swiftpm` it's a  package with "SwiftPM" name. 
Did I understood it correctly? 

We should generate 1 swift file for a root-package. That file will contain version info about root-package and it's dependencies.

mxcl commented:
When we say root package we mean the one swift built is executing against.

kostiakoval commented:
I've addressed 5 mentioned points:

- version should not be optional,
- keep `sha` optional
- modified as a `Bool`
- add label to all of the tuple values in version.
- Let’s add a versionString 

Few questions: 
- should we generate the same version data format for root-package and its dependencies? 
  now dependency package don't include `sha` and `modified`
  I think they should have same format.
- the url show the local path for a root-package. Should we use git origin if possible
- if there is no git repo for a root-package, should we still generate PackageVersions.swift file ?
  It is possible since we have local path as url and default version value

mxcl commented:
> should we generate the same version data format for root-package and its dependencies? now dependency package don't include sha and modified I think they should have same format.

My perspective is: yes.

> the url show the local path for a root-package. Should we use git origin if possible

Yes, use the origin if possible.

> if there is no git repo for a root-package, should we still generate PackageVersions.swift file ? It is possible since we have local path as url and default version value

I think we can be consistent and if there is no git data:

* `Version(0,0,0)`
* local path instead of origin

mxcl commented:
I'm happy to merge this if you are.

mxcl commented:
@swift-ci Please test

kostiakoval commented:
i'm in process implementing these changes. Will be done in 30 min

kostiakoval commented:
I don't feel this is optimal solution. I've tried to fix it but found more issue. We can merge it and improve it later if you like

mxcl commented:
We aim for more incremental development. It doesn't need to be perfect to be landed.

kostiakoval commented:
Agree. Let's wait for tests first

kostiakoval commented:
@mxcl Test are passing. Can we merge it. It would be easier to do smaller improvements than
